### PR TITLE
escrow offer: duration => max duration

### PR DESCRIPTION
**Pull request by kutoft**
high level problem:
offers in general are not working as well as we hoped they would. 
- mm offers are difficult as they are very time sensitive and so mm are reluctant to set a standing offer. The offers mm wants to support are also tied to deribit expiries instead of standard duration times, thus making them all very difficult to reuse.
- escrow offers are easier for suppliers to keep standing offers on, but since mm offers have such custom durations, matching an escrow offer to a mm offer becomes extremely difficult and unlikely.

approach:
This change focuses on helping make escrow offers more flexible to support the uniqueness of the mm offers that we see happening.
- by changing the escrow offer duration to a max duration, escrow offers are now capable of supporting the custom durations that occur from mm offers.
- this change means that the escrow will inherit the duration from the mm offer, ensuring that the durations always match.

note:
think of this PR as a conversation starter and assume there are plenty of discussions, reviews and code changes most likely needed in order to complete this effort (should we all agree on the direction).

### CODE REVIEW

File: src/EscrowSupplierNFT.sol [hunk 1]
```diff
@@ -198,16 +201,16 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
...
             interestAPR: SafeCast.toUint24(interestAPR),
             lateFeeAPR: SafeCast.toUint24(lateFeeAPR),
             minEscrow: minEscrow,
-            available: amount
+            available: amount,
+            maxDuration: maxDuration
         });
         asset.safeTransferFrom(msg.sender, address(this), amount);
         emit OfferCreated(
-            msg.sender, interestAPR, duration, gracePeriod, lateFeeAPR, amount, offerId, minEscrow
```
Comment (carlosdimatteo) [inline src/EscrowSupplierNFT.sol:210]: id replace duration with maxDuration to keep it in the event

File: src/LoansNFT.sol [hunk 2]
```diff
@@ -410,7 +410,7 @@ contract LoansNFT is ILoansNFT, BaseNFT {
 
         // handle optional escrow, must be done first, to use "supplier's" underlying in swap
         (EscrowSupplierNFT escrowNFT, uint escrowId) =
-            _conditionalOpenEscrow(usesEscrow, underlyingAmount, escrowOffer, escrowFees);
+            _conditionalOpenEscrow(usesEscrow, underlyingAmount, escrowOffer, escrowFees, providerOffer);
```
Comment (carlosdimatteo) [inline src/LoansNFT.sol:413]: probably would be better to pass the duration only instead of the whole object, specially cause its the only value being used

File: src/interfaces/IEscrowSupplierNFT.sol [hunk 3]
```diff
@@ -68,14 +69,14 @@ interface IEscrowSupplierNFT {
     event OfferCreated(
         address indexed supplier,
         uint indexed interestAPR,
-        uint indexed duration,
```
Comment (carlosdimatteo) [inline src/interfaces/IEscrowSupplierNFT.sol:71]: add max duration to event

File: src/LoansNFT.sol (hunk 2 above)
Comment (kutoft) [reply src/LoansNFT.sol:413]: my main reason for doing this here, is that _openLoan seemed to be a higher level function that didn't really deal in how something happened but just orchestrated making it happen. so I didn't want to pollute that function. but it doesnt really matter to me where I do it.

File: src/EscrowSupplierNFT.sol [hunk 4]
```diff
@@ -340,7 +362,7 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
         // The escrow funds are funds that have been escrowed in the ID being released ("O").
         // The offer is reduced (which is used to repay the previous supplier)
         // A new escrow ID is minted.
-        newEscrowId = _startEscrow(offerId, previousEscrow.escrowed, newFees, newLoanId);
+        newEscrowId = _startEscrow(offerId, previousEscrow.escrowed, newFees, newLoanId, previousEscrow.duration);
```
Comment (kutoft) [inline src/EscrowSupplierNFT.sol]: I am not confident we should reuse the old duration here. but not sure how else to handle it.

File: src/EscrowSupplierNFT.sol [hunk 5]
```diff
@@ -78,13 +78,15 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
         return nextTokenId;
     }
 
+
```
Comment (artdgn) [inline src/EscrowSupplierNFT.sol]: nit: forge fmt

File: src/EscrowSupplierNFT.sol [hunk 6]
```diff
@@ -172,22 +175,22 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
...
      */
     function createOffer(
         uint amount,
-        uint duration,
         uint interestAPR,
         uint gracePeriod,
         uint lateFeeAPR,
-        uint minEscrow
+        uint minEscrow,
+        uint maxDuration
```
Comment (artdgn) [inline src/EscrowSupplierNFT.sol]: info: better to keep the same order everywhere for easier visual comparison (types used, nothing skipped, etc) between inputs and assignments later.

File: src/EscrowSupplierNFT.sol [hunk 7]
```diff
@@ -198,7 +201,7 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
         offerId = nextOfferId++;
         offers[offerId] = OfferStored({
             supplier: msg.sender,
-            duration: SafeCast.toUint32(duration),
+            maxDuration: SafeCast.toUint32(maxDuration),
```
Comment (artdgn) [inline src/EscrowSupplierNFT.sol:206]: med: `minDuration` should also be added. Otherwise, if provider durations are allowed to be very short, there's a possible DoS of escrow offers by cycling them to withdrawals without escrow owners being compensated by enough fees (since fees depend on duration). An escrow owner may want to do it to remove competing offers.

While currently min provider duration is configured to 1mo, this code (Escrow) has no knowledge of it, and shouldn't rely on this assumption or be coupled to it. Also, it's possible for an escrow owner to want a higher minDuration than is configured at the time.

File: src/EscrowSupplierNFT.sol [hunk 8]
```diff
@@ -237,7 +240,22 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
...
+        emit OfferAmountUpdated(offerId, msg.sender, previousAmount, newAmount);
+    }
+    
+    /**
+     * @notice Updates the max duration of an existing offer.
+     * @dev Can only be called by the offer supplier
+     * @param offerId The ID of the offer to update
+     * @param newMaxDuration The new max duration in seconds
+     */
+    function updateOfferMaxDuration(uint offerId, uint newMaxDuration) external {
```
Comment (artdgn) [inline src/EscrowSupplierNFT.sol]: low: currently almost all stored values are immutable in all contracts, with the exception of offered amounts. Making the maxDuration modifiable adds another exception to this "values are immutable once set" rule, making it more difficult to ensure and review safe use of loaded / derived values. An example of this is the med issue with not validating duration during `switchEscrow` below (if it wasn't mutable, it would be low).

This also creates a slight (mostly negligible) race condition (frontrunning risk) between updating maxDuration and taking offers. It already exists for offer amounts, but is unavoidable.

It's better that if different max duration is needed, a new offer is created instead. I don't think there are any drawbacks to this (other than requiring two transactions instead of one, can be batched, is an edge case need anyway). Less code -> less mutability, less bugs, cheaper audits, less tests to write.

Additionally, this mutability creates a need to index another event and add correct handling of offer mutability in more places (backed, fronted, integrations)

File: src/EscrowSupplierNFT.sol (hunk 4 above)
Comment (artdgn) [reply src/EscrowSupplierNFT.sol]: I think `switchEscrow` should take a `newDuration` argument for the new position being created. It's used during a roll, and if provider durations aren't stable anymore, it makes sense that a new roll implementation will also be needed, and it will create a position with some `newDuration`.

In Loans, the `newDuration` can be loaded from the taker position using `_takerId(newLoadId)` after the roll.

If we add this now, it would work with current rolls, and with new rolls when it is needed, without having to redeploy escrows twice and migrate their users.

File: src/EscrowSupplierNFT.sol [hunk 9]
```diff
@@ -258,12 +276,16 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
...
      * @return escrowId The ID of the created escrow
      */
-    function startEscrow(uint offerId, uint escrowed, uint fees, uint loanId)
+    function startEscrow(uint offerId, uint escrowed, uint fees, uint loanId, uint duration)
         external
         returns (uint escrowId)
     {
+        // Check if duration is within the offer's max duration
+        Offer memory offer = getOffer(offerId);
+        require(duration <= offer.maxDuration, "escrow: duration exceeds offer's max duration");
```
Comment (artdgn) [inline src/EscrowSupplierNFT.sol]: med: this check needs to be in the validations in `_startEscrow()` instead. Currently it's not checked during `switchEscrow`. In this version of the code (with the ability to udpate maxDuration) it means a roll can ignore a new maxDuration value. If modifying duration is removed, in a future version of the code - if `switchEscrow` takes a `newDuration` as recommended (instead of using previous value) - it means that switchEscrow won't check the newDuration vs. maxDuration.

File: src/LoansNFT.sol (hunk 2 above)
Comment (artdgn) [reply src/LoansNFT.sol:413]: I think it's better as it is now (passing the whole offer struct into `_conditionalOpenEscrow`), as `_openLoan` is both a higher level flow, and is pretty complex already

File: src/LoansNFT.sol [hunk 10]
```diff
@@ -637,7 +637,7 @@ contract LoansNFT is ILoansNFT, BaseNFT {
 
     // ----- Conditional escrow mutative methods ----- //
 
-    function _conditionalOpenEscrow(bool usesEscrow, uint escrowed, EscrowOffer memory offer, uint fees)
+    function _conditionalOpenEscrow(bool usesEscrow, uint escrowed, EscrowOffer memory offer, uint fees, ProviderOffer memory providerOffer)
```
Comment (artdgn) [inline src/LoansNFT.sol]: nit: forge fmt

File: src/interfaces/IEscrowSupplierNFT.sol [hunk 11]
```diff
@@ -44,6 +44,7 @@ interface IEscrowSupplierNFT {
         uint escrowed;
         uint feesHeld;
         uint withdrawable;
+        uint32 duration; // duration for this escrow
```
Comment (artdgn) [inline src/interfaces/IEscrowSupplierNFT.sol]: info: can fit into the first slot (after `bool realeased`), saving gas.

File: src/EscrowSupplierNFT.sol [hunk 12]
```diff
@@ -172,22 +175,22 @@ contract EscrowSupplierNFT is IEscrowSupplierNFT, BaseNFT {
...
      * @notice Creates a new escrow offer
      * @param amount The offered amount
-     * @param duration The offer duration in seconds
      * @param interestAPR The annual interest rate in basis points. At most MAX_FEE_REFUND_BIPS
      * of the upfront interest can be refunded on cancellation. If interestAPR is 0, this
      * will have no effect, and allow free cancellations.
      * @param gracePeriod The maximum grace period duration in seconds
      * @param lateFeeAPR The annual late fee rate in basis points
      * @param minEscrow The minimum escrow amount. Protection from dust mints.
+     * @param maxDuration The duration in seconds for the escrow (calculated from loan expiration)
```
Comment (artdgn) [inline src/EscrowSupplierNFT.sol]: info: natspec outdated

**Review by artdgn — CHANGES_REQUESTED** [review]
A great first draft! Noted a few issues and suggestions for changes. 

Additional notes for further work on this:
- version strings need to be updated for the touched contracts
- changelog entries
- audits briefing (for any known issues)
- branch unit test coverage back to 100% if is reduced by the end
- deploy script changes
- fork test changes
